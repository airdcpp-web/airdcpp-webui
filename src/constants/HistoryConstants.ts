const MODULE_URL = 'histories';

export enum HistoryStringEnum {
  SEARCH = 'search_pattern',
  EXCLUDE = 'search_excluded',
  DOWNLOAD_DIR = 'download_target',
}

export enum HistoryEntryEnum {
  HUB = 'hub',
  PRIVATE_CHAT = 'private_chat',
  FILELIST = 'filelist',
}

export default {
  MODULE_URL: MODULE_URL,

  STRINGS_URL: MODULE_URL + '/strings',
  SESSIONS_URL: MODULE_URL + '/sessions',
};
