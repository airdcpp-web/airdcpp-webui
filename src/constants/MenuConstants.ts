const MODULE_URL = 'menus';
export default {
  MODULE_URL: MODULE_URL,

  ICON_TYPE_ID: 'semantic',

  QUEUE_BUNDLE: 'queue_bundle',
  QUEUE_FILE: 'queue_file',
  TRANSFER: 'transfer',
  SHARE_ROOT: 'share_root',
  FAVORITE_HUB: 'favorite_hub',
  USER: 'user',
  HINTED_USER: 'hinted_user',
  EXTENSION: 'extension',

  HUB_USER: 'hub_user',
  GROUPED_SEARCH_RESULT: 'grouped_search_result',
  FILELIST_ITEM: 'filelist_item',

  HUB_MESSAGE_HIGHLIGHT: 'hub_message_highlight',
  PRIVATE_CHAT_MESSAGE_HIGHLIGHT: 'private_chat_message_highlight',
};

export const MENU_SUPPORTS = ['urls', 'form'];
export const MENU_NESTING_THRESHOLD = 1;
