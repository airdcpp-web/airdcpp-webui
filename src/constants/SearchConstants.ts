const MODULE_URL = 'search';

export default {
  MODULE_URL: MODULE_URL,
  INSTANCES_URL: MODULE_URL,

  SEARCH_TYPES_URL: MODULE_URL + '/types',
  SEARCH_TYPES_UPDATED: 'search_types_updated',

  DEFAULT_SEARCH_TYPE: 'any',

  VIEW_ID: 'search',
};
