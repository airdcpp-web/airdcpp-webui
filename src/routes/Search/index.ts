export { default } from './components/Search';
