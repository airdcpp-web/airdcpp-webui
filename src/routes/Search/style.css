.search-container {
	width:100%;
	height: 50px;
}

#mobile-layout .search-container {
	margin-top: 10px;
}

.search-container .search-area {
	width:90%;
	max-width: 600px;
	margin-left: auto;
	margin-right: auto;
}

.search-area, .search-area .react-autosuggest__container {
	flex: 1;
}

.search-area, .search-area input {
	width: 100%;
}

.search-area button {
	display: inline-block !important;
}

.file-type-selector {
	float:left;
}

.search-layout {
	display: flex;
	flex-direction: column;
	height: 100%;
}

.search-layout .virtual-table {
	height: calc(100% - 50px);
}

@media only screen and (min-width: 767px) {
	.result.modal td {
		max-width: 250px;
	}
}

.result.modal .dupe-paths .path {
	overflow-x: hidden;
}

.result.modal .download.dropdown {
	max-width: 150px;
	margin-top: 10px;
}
