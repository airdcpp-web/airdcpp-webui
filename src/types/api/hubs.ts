import { ChatMessageCounts, EncryptionInfo } from './common';
import { ShareProfileBasic } from './shareprofiles';

export const enum HubConnectStateEnum {
  REDIRECT = 'redirect',
  CONNECTING = 'connecting',
  PASSWORD = 'password',
  CONNECTED = 'connected',
  DISCONNECTED = 'disconnected',
  KEYPRINT_ERROR = 'keyprint_mismatch',
}

export interface HubConnectState {
  id: HubConnectStateEnum;
  str: string;
  data?: {
    hub_url?: string;
  };
}

export interface HubSettings {
  use_main_chat_notify: boolean;
  show_joins: boolean;
  fav_show_joins: boolean;
  nick: string;
}

export interface Hub {
  id: number;
  hub_url: string;
  connect_state: HubConnectState;
  identity: {
    name: string;
    description: string;
  };
  share_profile?: ShareProfileBasic;
  favorite_hub?: number | null;
  message_counts: ChatMessageCounts;
  encryption: EncryptionInfo | null;
  settings: HubSettings;
}

export interface HubCounts {
  share_size: number;
  user_count: number;
}
