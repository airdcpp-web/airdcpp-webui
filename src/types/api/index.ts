export * from './common';
export * from './table';

export * from './connectivity';
export * from './extensions';
export * from './favorite-directories';
export * from './favorite-hubs';
export * from './filelists';
export * from './filesystem';
export * from './hash';
export * from './history';
export * from './hubs';
export * from './menu';
export * from './private-chat';
export * from './queue';
export * from './search';
export * from './sessions';
export * from './settings';
export * from './share';
export * from './shareprofiles';
export * from './system';
export * from './table';
export * from './transfers';
export * from './viewed-files';
export * from './web-users';