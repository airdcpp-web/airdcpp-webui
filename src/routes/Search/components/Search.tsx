import React from 'react';

import SocketService from 'services/SocketService';

import { HistoryStringEnum } from 'constants/HistoryConstants';
import SearchConstants from 'constants/SearchConstants';

import History from 'utils/History';
import HistoryInput from 'components/autosuggest/HistoryInput';

import OfflineHubMessageDecorator from 'decorators/OfflineHubMessageDecorator';

import Button from 'components/semantic/Button';

import '../style.css';
import ResultTable from 'routes/Search/components/ResultTable';
import { RouteComponentProps } from 'react-router-dom';

import * as API from 'types/api';


const SEARCH_PERIOD = 4000;

interface SearchProps extends RouteComponentProps<{}> {

}

class Search extends React.Component<SearchProps> {
  state = {
    searchString: '',
    running: false
  };

  _searchTimeout: NodeJS.Timer;

  componentDidMount() {
    this.checkLocationState(this.props);
  }

  componentDidUpdate(prevProps: SearchProps) {
    this.checkLocationState(this.props);
  }

  checkLocationState = (props: SearchProps) => {
    const { state } = props.location;
    if (state && state.searchString && state.searchString !== this.state.searchString) {
      this.search(state.searchString);

      // Avoid searching for it again
      History.replace({
        pathname: props.location.pathname,
      });
    }
  }

  search = (searchString: string) => {
    console.log('Searching');

    clearTimeout(this._searchTimeout);

    SocketService.post(SearchConstants.HUB_SEARCH_URL, {
      query: {
        pattern: searchString,
      },
      priority: API.PriorityEnum.HIGH,
    })
      .then(this.onSearchPosted)
      .catch((error: string) => 
        console.error('Failed to post search: ' + error)
      );

    this.setState({
      searchString,
      running: true 
    });
  }

  onSearchPosted = (data: API.SearchResponse) => {
    this._searchTimeout = setTimeout(
      () => {
        this.setState({ 
          running: false,
        });
      }, 
      data.queue_time + SEARCH_PERIOD
    );
  }

  render() {
    const { searchString, running } = this.state;
    var searchDefault = searchString;
    var url = new URL(window.location.href);
    var urlSearch = url.searchParams.get("query");
    if (urlSearch!=null) {
      searchDefault = urlSearch;
      console.log('Search term from url: ' + searchDefault);
    }
    return (
      <OfflineHubMessageDecorator 
        offlineMessage="You must to be connected to at least one hub in order to perform searches"
      >
        <div className="search-layout">
          <div className="search-container">
            <div className="search-area">
              <HistoryInput 
                historyId={ HistoryStringEnum.SEARCH } 
                submitHandler={ this.search } 
                disabled={ running }
                defaultValue={ searchDefault }
                placeholder="Enter search string..."
                button={ 
                  <Button
                    icon="search icon"
                    caption="Search"
                    loading={ running }
                  />
                }
              />
            </div>
          </div>
          <ResultTable 
            searchString={ searchString } 
            running={ running }
          />
        </div>
      </OfflineHubMessageDecorator>
    );
  }
}

export default Search;
