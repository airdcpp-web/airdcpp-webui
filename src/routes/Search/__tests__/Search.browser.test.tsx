import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';

import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import { installTableMocks } from '@/tests/mocks/mock-table';

import { waitFor } from '@testing-library/dom';
import {
  clickButton,
  clickIconButton,
  navigateToUrl,
  waitExpectRequestToMatchSnapshot,
  waitForLoader,
  waitForUrl,
} from '@/tests/helpers/test-helpers';

import Search from '../components/Search';
import SearchConstants from '@/constants/SearchConstants';
import {
  GroupedSearchResultChildrenResponse,
  GroupedSearchResultDirectoryResponse,
  GroupedSearchResultFileResponse,
  GroupedSearchResultsListResponse,
  SearchInstanceHubSearchPostResponse,
  SearchInstanceListResponse,
  SearchInstanceNewResponse,
  SearchTypesListResponse,
} from '@/tests/mocks/api/search';
import HistoryConstants, { HistoryStringEnum } from '@/constants/HistoryConstants';
import {
  setInputFieldValuesByName,
  setInputFieldValuesByPlaceholder,
  setupUserEvent,
  toggleCheckboxValue,
} from '@/tests/helpers/test-form-helpers';
import { HubADC1, HubADC2, HubADC3, HubNMDC1 } from '@/tests/mocks/api/hubs';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
import { clickMenuItem, openMenu } from '@/tests/helpers/test-menu-helpers';
import { installActionMenuMocks } from '@/components/action-menu/__tests__/test-action-menu-helpers';
import MenuConstants from '@/constants/MenuConstants';

describe('Search layout', () => {
  let server: MockServer;

  beforeEach(() => {
    server =
      getMockServer(/*{
      loggerOptions: {
        logLevel: 'verbose',
      },
    }*/);
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(
      server /*, {
      socketOptions: {
        logLevel: 'verbose',
      },
    }*/,
    );

    // Tables
    const resultTableMocks = installTableMocks([], {
      // The view should be initially empty
      server,
      moduleUrl: SearchConstants.MODULE_URL,
      viewName: SearchConstants.VIEW_ID,
      entityId: SearchInstanceNewResponse.id,
    });

    // Search instances
    server.addRequestHandler(
      'GET',
      SearchConstants.INSTANCES_URL,
      SearchInstanceListResponse,
    );

    const onCreateInstance = vi.fn();
    server.addRequestHandler(
      'POST',
      SearchConstants.INSTANCES_URL,
      SearchInstanceNewResponse,
      onCreateInstance,
    );

    server.addRequestHandler(
      'GET',
      `${SearchConstants.INSTANCES_URL}/${SearchInstanceNewResponse.id}`,
      SearchInstanceNewResponse,
    );

    // Instance
    const onPostHubSearch = vi.fn();
    server.addRequestHandler(
      'POST',
      `${SearchConstants.INSTANCES_URL}/${SearchInstanceNewResponse.id}/hub_search`,
      SearchInstanceHubSearchPostResponse,
      onPostHubSearch,
    );

    // Results
    installActionMenuMocks(MenuConstants.GROUPED_SEARCH_RESULT, [], server);

    server.addRequestHandler(
      'GET',
      // eslint-disable-next-line max-len
      `${SearchConstants.INSTANCES_URL}/${SearchInstanceNewResponse.id}/results/${GroupedSearchResultDirectoryResponse.id}/children`,
      GroupedSearchResultChildrenResponse,
    );

    server.addRequestHandler(
      'GET',
      // eslint-disable-next-line max-len
      `${SearchConstants.INSTANCES_URL}/${SearchInstanceNewResponse.id}/results/${GroupedSearchResultDirectoryResponse.id}`,
      GroupedSearchResultDirectoryResponse,
    );

    // Histories
    const onAddHistory = vi.fn();
    server.addRequestHandler(
      'POST',
      `${HistoryConstants.STRINGS_URL}/${HistoryStringEnum.SEARCH}`,
      undefined,
      onAddHistory,
    );

    server.addRequestHandler(
      'GET',
      `${HistoryConstants.STRINGS_URL}/${HistoryStringEnum.SEARCH}`,
      ['search string old', 'C5HMPSV2L7EW4GZBTG6H4MUPKGQLAZUAGZNACHY'],
    );

    // Options
    server.addRequestHandler(
      'GET',
      SearchConstants.SEARCH_TYPES_URL,
      SearchTypesListResponse,
    );

    return {
      commonData,
      server,
      resultTableMocks,
      onCreateInstance,
      onAddHistory,
      onPostHubSearch,
    };
  };

  const renderLayout = async () => {
    const { commonData, server, ...other } = await getSocket();

    const SearchLayoutTest = () => {
      return <Search />;
    };

    const routes = [
      {
        index: true,
        Component: () => <div>Index page</div>,
      },
      {
        path: '/search/*',
        Component: SearchLayoutTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/search'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    const userEvent = setupUserEvent();
    return { ...renderData, ...other, ...commonData, userEvent };
  };

  test('should render', async () => {
    const renderData = await renderLayout();
    const {
      getByRole,
      queryByText,
      getByPlaceholderText,
      getByText,

      sessionStore,
      router,
      resultTableMocks,

      onCreateInstance,
      onPostHubSearch,
      onAddHistory,
    } = renderData;

    await waitFor(() => expect(onCreateInstance).toHaveBeenCalled());

    // Search shouldn't be available if we aren't connected to any hub
    await waitFor(() => expect(queryByText('No online hubs')).toBeTruthy());

    // Create a session, the search input should appear
    sessionStore.getState().hubs.init([HubADC1]);

    await waitFor(() =>
      expect(getByPlaceholderText('Enter search string...')).toBeTruthy(),
    );

    // Perform our search
    await setInputFieldValuesByPlaceholder(renderData, {
      'Enter search string...': 'ubuntu',
    });

    clickButton('Search', getByRole);

    await waitExpectRequestToMatchSnapshot(onPostHubSearch);
    await waitFor(() => expect(onAddHistory).toHaveBeenCalled());

    // Wait for the search results to be displayed
    const button = getByRole('button', { name: 'Search' });
    await waitFor(() => expect(button).toBeDisabled());

    // Send the results
    resultTableMocks.mockTableManager.setItems(GroupedSearchResultsListResponse);

    // The table should appear
    await waitFor(() => expect(getByRole('grid')).toBeTruthy());

    const groupedResult = GroupedSearchResultDirectoryResponse;
    await waitFor(() => expect(queryByText(groupedResult.name)).toBeTruthy());

    // Check dupe classes
    const nonDupeResult = getByText(GroupedSearchResultDirectoryResponse.name).closest(
      `.name`,
    );
    expect(nonDupeResult?.classList.contains('dupe')).toBeFalsy();

    const dupeResult = getByText(GroupedSearchResultFileResponse.name).closest(`.name`);
    expect(dupeResult?.classList.contains('dupe')).toBeTruthy();

    await navigateToUrl('/', router);
  });

  test('should open result details dialog', async () => {
    const renderData = await renderLayout();
    const {
      getByRole,
      queryByText,

      sessionStore,
      router,
      resultTableMocks,
    } = renderData;

    sessionStore.getState().hubs.init([HubADC1, HubADC2, HubADC3]);
    resultTableMocks.mockTableManager.setItems(GroupedSearchResultsListResponse);

    await waitFor(() => expect(getByRole('grid')).toBeTruthy());

    const result = GroupedSearchResultDirectoryResponse;
    await waitFor(() => expect(queryByText(result.name)).toBeTruthy());

    // Open the result dialog
    const contentCaption = result.name;
    await openMenu(contentCaption, renderData);
    await clickMenuItem('Result details', renderData);

    await waitFor(() => expect(getByRole('dialog')).toBeTruthy());

    // Close it
    clickButton('Close', getByRole);
    await waitForUrl('/search', router);

    await navigateToUrl('/', router);
  });

  test('should handle custom search options', async () => {
    const renderData = await renderLayout();
    const {
      getByRole,
      getByPlaceholderText,
      queryByRole,

      sessionStore,
      router,

      onPostHubSearch,
      userEvent,
    } = renderData;

    // Create a session, the search input should appear
    sessionStore.getState().hubs.init([HubADC1, HubADC2, HubADC3, HubNMDC1]);

    await waitFor(() =>
      expect(getByPlaceholderText('Enter search string...')).toBeTruthy(),
    );

    // Open search options
    await waitFor(() =>
      expect(getByRole('button', { name: 'Open search options' })).toBeTruthy(),
    );
    await clickIconButton('Open search options', renderData);

    await waitFor(() => expect(getByRole('form')).toBeTruthy());
    await waitForLoader(queryByRole);

    // Select type
    await openMenu('Any', renderData);
    await clickMenuItem('File', renderData);

    // Unselect a hub
    await userEvent.click(getByRole('button', { name: 'Hubs' }));
    await toggleCheckboxValue(HubADC3.identity.name, renderData);

    // Set size
    await userEvent.click(getByRole('button', { name: 'Size limits' }));
    await setInputFieldValuesByName(renderData, {
      'Minimum size': '1000',
      'Maximum size': '5000',
    });

    // Perform our search
    await setInputFieldValuesByPlaceholder(renderData, {
      'Enter search string...': 'ubuntu',
    });

    clickButton('Search', getByRole);

    await waitExpectRequestToMatchSnapshot(onPostHubSearch);

    await navigateToUrl('/', router);
  });
});
