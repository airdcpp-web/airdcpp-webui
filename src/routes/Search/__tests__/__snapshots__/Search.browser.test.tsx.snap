// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search layout > should handle custom search options 1`] = `
{
  "data": {
    "hub_urls": [
      "localhost.nmdc",
      "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
    ],
    "priority": 5,
    "query": {
      "file_type": "file",
      "max_size": 5242880000,
      "min_size": 1048576000,
      "pattern": "ubuntu",
    },
  },
  "method": "POST",
  "path": "search/3/hub_search",
}
`;

exports[`Search layout > should render 1`] = `
{
  "data": {
    "priority": 5,
    "query": {
      "pattern": "ubuntu",
    },
  },
  "method": "POST",
  "path": "search/3/hub_search",
}
`;
